File: frontend/types/document.ts
export type DocumentStatusValue = "pending" | "processing" | "ready" | "indexed" | "error";

// pending: not yet started · generating: embeddings running · complete: searchable ·
// error: permanently failed after retries (status stays "ready", not "error")
export type EmbeddingStatusValue = "pending" | "generating" | "complete" | "error";

export type ChunkType = "paragraph" | "table" | "list" | "header+content";

export interface Document {
  id: string;
  filename: string;
  display_name: string | null;
  mime_type: string | null;
  project_id: string | null;
  file_key: string;
  file_size_bytes: number | null;
  status: DocumentStatusValue;
  embedding_status: EmbeddingStatusValue;
  page_count: number | null;
  word_count: number | null;
  chunk_count: number | null;
  extraction_method: string | null;
  summary: string | null;
  error_message: string | null;
  processing_started_at: string | null;
  processing_completed_at: string | null;
  created_at: string;
  updated_at: string;
  uploaded_by: string;
  file_type_label: string;
}

export interface DocumentDetail extends Document {
  uploaded_by_name: string | null;
  extracted_text_preview: string | null;
  embedded_chunk_count: number | null;
  embedding_model: string | null;
}

export interface DocumentStatusPoll {
  id: string;
  status: DocumentStatusValue;
  embedding_status: EmbeddingStatusValue;
  progress_percent: number;
  error_message: string | null;
  word_count: number | null;
  page_count: number | null;
  chunk_count: number | null;
}

export interface Chunk {
  id: string;
  document_id: string;
  chunk_index: number;
  content: string;
  chunk_type: ChunkType | null;
  page_number: number | null;
  token_count: number | null;
  created_at: string;
}

File: backend/mednotebook_backend/models/document.py
import enum
import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from ..database import Base


class DocumentStatus(str, enum.Enum):
    pending = "pending"
    processing = "processing"
    ready = "ready"
    # Text extracted, chunked, and every chunk has an embedding — searchable.
    indexed = "indexed"
    error = "error"


class Document(Base):
    __tablename__ = "documents"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="SET NULL"), nullable=True, index=True
    )
    uploaded_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    filename: Mapped[str] = mapped_column(String(255), nullable=False)
    display_name: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    file_key: Mapped[str] = mapped_column(String(500), nullable=False)
    file_size_bytes: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    mime_type: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    status: Mapped[DocumentStatus] = mapped_column(
        Enum(DocumentStatus), default=DocumentStatus.pending, nullable=False
    )
    page_count: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    word_count: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    chunk_count: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    extraction_method: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    # Tracks embedding progress specifically (pending/generating/complete/error) —
    # kept in sync with `status` by the Celery pipeline but distinct because it
    # captures a "generating" state that `status` (ready/processing/indexed) doesn't.
    embedding_status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="pending")
    summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    processing_started_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    processing_completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    project: Mapped["Project"] = relationship("Project", back_populates="documents")
    uploader: Mapped["User"] = relationship("User", back_populates="uploaded_documents")
    chunks: Mapped[list["DocumentChunk"]] = relationship(
        "DocumentChunk", back_populates="document", cascade="all, delete-orphan"
    )
    embedding_costs: Mapped[list["EmbeddingCost"]] = relationship(
        "EmbeddingCost", back_populates="document", cascade="all, delete-orphan"
    )
